Pull portability macros from feature/nvcxx_compatibility

wmaxey opened the pull request.

wmaxey commented:
APIs provided by this:

* `NV_IF_TARGET(cond, t, ...)`
* `NV_DISPATCH_TARGET(...)`

Internally used *ONLY* to enable/disable headers and features:
`_NV_TARGET_MINIMUM_VAL` - Provides a preprocessor time integer value representing the lowest SM architecture supplied at compile time. This behaves very differently on NVCC and NVC++.

dkolsen-pgi commented on include/nv/detail/__target_macros:
```diff
@@ -0,0 +1,395 @@
+#  define _NV_TARGET_VAL_SM_80 nv::target::sm_80
+#  define _NV_TARGET_VAL_SM_86 nv::target::sm_86
+
+#  define _NV_TARGET_VAL nv::target::device
```
`nv::target::device` is no longer defined.  Which makes me think that `_NV_TARGET_VAL` is no longer used in the non-NVCC branch.  `_NV_TARGET_VAL` should be removed from the NVC++ section.

dkolsen-pgi commented on include/nv/detail/__target_macros:
```diff
@@ -0,0 +1,395 @@
+#  define _NV_TARGET___NV_ANY_TARGET (nv::target::is_host || nv::target::is_device)
+#  define _NV_TARGET___NV_NO_TARGET !_NV_TARGET___NV_ANY_TARGET
+
+#if defined(__NVCOMPILER_CUDA_ARCH__)
```
`__NVCOMPILER_CUDA_ARCH__` will go away once Allison finishes refactoring CUB.  It shouldn't be referenced at all by `<nv/target>`.  I think `_NV_TARGET_DEF` is not necessary.  `_NV_TARGET_MINIMUM_VAL` should be kept, but on its own and not within a `#if defined(__NVCOMPILER_CUDA_ARCH__)` block.

wmaxey replied:
I don't see it being used in the code base at this point

dkolsen-pgi commented on include/nv/target:
```diff
@@ -0,0 +1,154 @@
+
+#if defined(__NVCC__) || defined(__CUDACC_RTC__)
+#  define _NV_COMPILER_NVCC
+#elif defined(__PGIC__)
```
`#elif defined(__NVCOMPILER)`

`__PGIC__` is old name that shouldn't be used anymore.

dkolsen-pgi commented on include/nv/detail/__target_macros:
```diff
@@ -0,0 +1,395 @@
+#  define _NV_TARGET_VAL_SM_80 800
+#  define _NV_TARGET_VAL_SM_86 860
+
+// There is a case where __CUDA_ARCH__ is 0 while undefined that can cause host code paths to emit a device code dispatch entry
```
Get rid of this comment.  It doesn't apply here.

dkolsen-pgi commented on .upstream-tests/test/cuda/test_platform.pass.cpp:
```diff
@@ -0,0 +1,234 @@
+
+#if defined(__NVCC__)
+#  define TEST_NVCC
+#elif defined(__PGIC__)
```
`#elif defined(__NVCOMPILER)`

brycelelbach commented:
Can we add `nv::target::any_target` and `nv::target::no_target`.

wmaxey commented:
I removed a couple commits that were redundant. I've implemented `nv::target::any_target` and `no_target`

@brycelelbach or @dkolsen-pgi for re-review.

dkolsen-pgi commented on include/nv/target:
```diff
@@ -0,0 +1,158 @@
+    constexpr target_description any_target =
+      target_description(detail::all_hosts | detail::all_devices);
+    constexpr target_description no_target =
+      target_description(detail::all_hosts & detail::all_devices);
```
This should just be `no_target = target_description(0)`

wmaxey commented:
Local tests for all supported host compilers are clear. GCC, clang, and MSVC happily produce host only code.
